### PR TITLE
Add two new commands to mkcloud (cloudsnapshot)

### PR DESCRIPTION
**Pull request by nicolasbock**
It can be desireble to preserve the state of the full cloud for debugging and
development purposes. This patch adds two new commands to mkcloud,
createcloudsnapshot, and restorefromcloudsnapshot, which snapshot all disks in
the cloud.

### CODE REVIEW

**Comment by nicolasbock** [comment]
I have used the two new commands now several times and they seem to indeed do what they are supposed to.

**Comment by nicolasbock** [comment]
@aplanas Could you review?

**Comment by aplanas** [comment]
+1 very appreciated this feature

**Comment by bmwiedemann** [comment]
maybe think about ways to make it DRY = reduce duplication with createadminsnapshot and restore

**Comment by jsuchome** [comment]
+1

File: scripts/mkcloud [hunk 1]
```diff
@@ -539,12 +539,32 @@ function setupadmin()
...
 }
 
+function wait_for_node_shutdown()
+{
+    wait_for 150 1 "virsh domstate $1 | grep shut.off" '$1 node to shut down'
+}
+
+function remove_snapshot_volume()
+{
+    lvremove -f $1
```
Comment (jdsn) [inline scripts/mkcloud]: This could remove any LV. How about checking if the volume really is a snapshot volume first?

File: scripts/mkcloud [hunk 2]
```diff
@@ -1011,25 +1063,37 @@ Steps:
     cleanup:        kill all running VMs, zero out boot sectors of all LVM volumes
     prepare:        create LVM volumes, setup libvirt networks
     setupadmin:     create the admin node and install the cloud product
-    prepareinstcrowbar: add repos and install crowbar packages
+    prepareinstcrowbar:
+                    add repos and install crowbar packages
```
Comment (jdsn) [inline scripts/mkcloud]: many unrelated changes in this commit - even unrelated to the PR
please split it off into separate PR

File: scripts/mkcloud [hunk 3]
```diff
@@ -1011,25 +1063,37 @@ Steps:
...
-    createadminsnapshot: create snapshot of admin node disk
-    restoreadminfromsnapshot: restore admin node disk from snapshot
+    createadminsnapshot:
+                    create snapshot of admin node disk
+    restoreadminfromsnapshot:
+                    restore admin node disk from snapshot
+    createcloudsnapshot:
+                    create a snapshot of all nodes
+    restorecloudfromsnapshot:
+                    restore all nodes from snapshot
```
Comment (jdsn) [inline scripts/mkcloud]: these are fine however

File: scripts/mkcloud [hunk 4]
```diff
@@ -540,12 +540,36 @@ function setupadmin()
...
 }
 
+function wait_for_node_shutdown()
+{
+    wait_for 150 1 "virsh domstate $1 | grep shut.off" '$1 node to shut down'
+}
+
+function remove_snapshot_volume()
+{
+    if [[ "${1}" =~ \.snap$ ]]; then
```
Comment (nicolasbock) [inline scripts/mkcloud]: We could also move the whole naming convention into the function, i.e. we call the `.*snapshot.*` functions with a bare volume name and then append `.snap` to the names as needed.

**Comment by MaximilianMeister** [comment]
+1

File: scripts/mkcloud (hunk 4 above)
Comment (jdsn) [inline scripts/mkcloud]: yes, that might make more sense

and btw. `if [[ $1 =~ \.snap$ ]]` works without quotes

File: scripts/mkcloud [hunk 5]
```diff
@@ -540,12 +540,36 @@ function setupadmin()
...
+
 function createadminsnapshot()
 {
     virsh shutdown $cloud-admin
-    wait_for 150 1 "virsh domstate $cloud-admin|grep shut.off" 'admin node to shut down'
-    lvremove -f $admin_node_disk.snap
-    safely lvcreate -l100%ORIGIN -s -n $cloud.admin.snap $admin_node_disk
+    wait_for_node_shutdown $cloud-admin
+    remove_snapshot_volume $admin_node_disk.snap
+    create_snapshot_volume $admin_node_disk.snap $admin_node_disk
```
Comment (jdsn) [inline scripts/mkcloud]: then also move the naming of the snapshot creation into the function

File: scripts/mkcloud (hunk 4 above)
Comment (jdsn) [inline scripts/mkcloud]: > We could also move the whole naming convention into the function, i.e. we call the .*snapshot.* functions with a bare volume name and then append .snap to the names as needed.

Hint for the proposed change: instead of the `if else` block just add a `safely` in front of the command you want to secure. That does (almost) the same but is shorter and more readable.

**Comment by jdsn** [comment]
:-1: for now, see comments

File: scripts/mkcloud (hunk 4 above)
Comment (nicolasbock) [inline scripts/mkcloud]: > and btw. if [[ $1 =~ \.snap$ ]] works without quotes

Only if the path `$1` does not contain whitespace.

Comment (jdsn) [inline scripts/mkcloud]: > Only if the path $1 does not contain whitespace.

Nope.
Double brackets `[[ ]]` are intelligent, they know that this is just one variable, so it also works with whitespace and even with empty string. The goal of using `[[` is to reduce code and make it more readable. So we should use the features it offers.

See:

```
foo="foo bar"
[[ $foo =~ foo ]] && echo works
[[ $foo =~ bar ]] && echo works
[[ $foo =~ "foo bar" ]] && echo works # only here you need to quote the string, but not the variable
```

Comment (nicolasbock) [inline scripts/mkcloud]: I didn't realize that. Very cool!

**Comment by nicolasbock** [comment]
@jdsn Could you please have another look?

File: scripts/mkcloud [hunk 6]
```diff
@@ -540,24 +540,84 @@ function setupadmin()
     fi
 }
 
+function shutdown_node() {
+    virsh shutdown $1
+}
```
Comment (jdsn) [inline scripts/mkcloud]: Is that worth a function?

File: scripts/mkcloud [hunk 7]
```diff
@@ -540,24 +540,84 @@ function setupadmin()
...
+
+function destroy_node() {
+    virsh destroy $1
+}
```
Comment (jdsn) [inline scripts/mkcloud]: Is that worth a function?

File: scripts/mkcloud (hunk 6 above)
Comment (nicolasbock) [inline scripts/mkcloud]: Well, it's true that it's a very short function :smirk: I figured though that if we ever want to change the virtulization approach we just have to change the function.

File: scripts/mkcloud (hunk 7 above)
Comment (nicolasbock) [inline scripts/mkcloud]: Same argument here. This isolates the virtualization details into these two functions.

File: scripts/mkcloud (hunk 6 above)
Comment (jdsn) [inline scripts/mkcloud]: Well ok, fine.
( sidenote: I expect that we will no longer have this part in bash code when we add other virtualization approaches )

Comment (jdsn) [inline scripts/mkcloud]: btw. `virsh start $cloud-admin` is still un-functionized a few lines below
So when we support a cloud in a cloud we will have to touch this code anyway. So again, I don't really see the point of growing the number of functions just for the sake of having a different name.

Comment (nicolasbock) [inline scripts/mkcloud]: I'll revert those calls and remove the function. That will make the `virsh start` calls consistent across the code.

**Comment by nicolasbock** [comment]
@jdsn I removed the two `virsh` functions again. Please have another look.

**Comment by jdsn** [comment]
:+1:

**Comment by MaximilianMeister** [comment]
nice and helpful feature! thanks for working on this

**Comment by aspiers** [comment]
Same as what @MaximilianMeister said :)
